tools/clang/tools/dxcompiler/dxcpdbutils.cpp

```diff
@@ -65,21 +70,139 @@ static HRESULT CopyWstringToBSTR(const std::wstring &str, BSTR *pResult) {
       return false;
     }
   }
   return true;
 }
 
+static std::vector<std::wstring> ComputeFlagsBasedOnArgs(ArrayRef<std::wstring> args) {
+  std::vector<std::wstring> flags;
+  // Flags - which exclude entry point, target profile, and defines
+  for (unsigned i = 0; i < args.size(); i++) {
+    const std::wstring &arg = args[i];
+    bool skip_another_arg = false;
+    if (ShouldIncludeInFlags(arg, &skip_another_arg)) {
+      flags.push_back(arg);
+    }
+    if (skip_another_arg)
+      i++;
+  }
+  return flags;
+}
+
+struct DxcPdbVersionInfo : public IDxcVersionInfo2 {
+private:
+  DXC_MICROCOM_TM_REF_FIELDS()
+
+public:
+  DXC_MICROCOM_TM_ADDREF_RELEASE_IMPL()
+  DXC_MICROCOM_TM_ALLOC(DxcPdbVersionInfo)
+
+  DxcPdbVersionInfo(IMalloc *pMalloc) : m_dwRef(0), m_pMalloc(pMalloc) {}
+
+  hlsl::DxilCompilerVersion m_Version = {};
+
+  HRESULT STDMETHODCALLTYPE QueryInterface(REFIID iid, void **ppvObject) override {
+    return DoBasicQueryInterface<IDxcVersionInfo, IDxcVersionInfo2>(this, iid, ppvObject);
+  }
+
+  virtual HRESULT STDMETHODCALLTYPE GetVersion(_Out_ UINT32 *pMajor, _Out_ UINT32 *pMinor) override {
+    if (!pMajor || !pMinor)
+      return E_POINTER;
+    *pMajor = m_Version.Major;
+    *pMinor = m_Version.Minor;
+    return S_OK;
+  }
+
+  virtual HRESULT STDMETHODCALLTYPE GetFlags(_Out_ UINT32 *pFlags) {
+    if (!pFlags) return E_POINTER;
+    *pFlags = m_Version.VersionFlags;
+    return S_OK;
+  }
+
+  virtual HRESULT STDMETHODCALLTYPE GetCommitInfo(_Out_ UINT32 *pCommitCount, _Out_ char **pCommitHash) {
+    if (!pCommitHash)
+      return E_POINTER;
+
+    *pCommitHash = nullptr;
+
+    char *const hash = (char *)CoTaskMemAlloc(sizeof(m_Version.CommitSha) + 1);
+    if (hash == nullptr)
+      return E_OUTOFMEMORY;
+    std::memcpy(hash, m_Version.CommitSha, sizeof(m_Version.CommitSha));
+    hash[sizeof(m_Version.CommitSha)] = 0;
+
+    *pCommitHash = hash;
+    *pCommitCount = m_Version.CommitCount;
+
+    return S_OK;
+  }
+};
+
+struct PdbRecompilerIncludeHandler : public IDxcIncludeHandler {
+  private:
+  DXC_MICROCOM_TM_REF_FIELDS()
+
+  public:
+  DXC_MICROCOM_TM_ADDREF_RELEASE_IMPL()
+  DXC_MICROCOM_TM_ALLOC(PdbRecompilerIncludeHandler)
+
+  PdbRecompilerIncludeHandler(IMalloc *pMalloc) : m_dwRef(0), m_pMalloc(pMalloc) {}
+
+  std::unordered_map< std::wstring, unsigned > m_FileMap;
+  IDxcPdbUtils *m_pPdbUtils = nullptr;
+
+  UINT32 m_Major = 0;
+  UINT32 m_Minor = 0;
+  std::string m_Hash;
+  UINT32 m_Flags = 0;
+  UINT32 m_CommitCount = 0;
+
+  HRESULT STDMETHODCALLTYPE QueryInterface(REFIID iid, void **ppvObject) override {
+    return DoBasicQueryInterface<IDxcIncludeHandler>(this, iid, ppvObject);
+  }
+
+  virtual HRESULT STDMETHODCALLTYPE LoadSource(
+    _In_z_ LPCWSTR pFilename,                                 // Candidate filename.
+    _COM_Outptr_result_maybenull_ IDxcBlob **ppIncludeSource  // Resultant source object for included file, nullptr if not found.
```

Comment:
COM rules indicate that all out-params should be initialized to a known value (i.e. set it to nullptr) if the method returns an error.